### PR TITLE
Bug fixes for Project dependency tree nodes hierarchy

### PR DESCRIPTION
**Pull request by abpiskunov**
There were several issues with Project nodes hierarchy:
- when A -> B and B is several levels above A, like ..\..\..\b.csproj, progression code cut node id string when coverted it to AbsoluteOr Relative Uri. We need to UrlEncode/Decode id string for nodes.
- other issue is we need to use Project B dependency providers when querying hierarchy nodes under project A Dependencies to show hierarchy correctly: A -> B (and B has packages specific to B) . Before we used providers from project A and they did not know about nodes specific to B and worked only when A also had same packages as B.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs [hunk 1]
```diff
@@ -237,13 +237,12 @@ public override IProjectTree FindByPath(IProjectTree root, string path)
...
-            // Note: all dependency nodes file path starts with file:/// to make sure we have 
-            // valid absolute path everytime.
-            if (!path.StartsWith("file:///"))
-            {
-                // just in case if given path is not in uri format
-                path = "file:///" + path.Trim('/');
-            }
+            // Note: all dependency nodes file path starts with file%3a (encoded 'file:'). 
+            // So if given path is not starting with file%3a, we are not responsible for it.
+            //if (!path.StartsWith("file%3a"))
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs]: this comment should be removed

**Comment by abpiskunov** [comment]
@dotnet/project-system

**Comment by abpiskunov** [comment]
Yes, right , will remove the tuple.


On Dec 29, 2016, at 4:07 PM, Fred Silberberg <notifications@github.com<mailto:notifications@github.com>> wrote:


@333fred commented on this pull request.

________________________________

In src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesGraphProvider.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Froslyn-project-system%2Fpull%2F1055%23pullrequestreview-14723124&data=02%7C01%7Canton.piskunov%40microsoft.com%7Cad218b905de64d043d9208d43047e4c2%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636186532712744421&sdata=eGkf%2BgTMt9hLTiaVd7TWYlwNMXfV%2BUiWVk5h57G8ozw%3D&reserved=0>:

>          {
-            IProjectDependenciesSubTreeProvider subTreeProvider;
-            var node = inputGraphNode.GetValue<IDependencyNode>(
-                                DependenciesGraphSchema.DependencyNodeProperty);
+            (IDependencyNode node, IProjectDependenciesSubTreeProvider subTreeProvider) = (null, null);


Why use tuples here? You're needlessly creating a ValueTuple and destructuring it for no purpose. Just make two variables.

-

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyNodeId.cs [hunk 2]
```diff
@@ -96,9 +108,12 @@ public override string ToString()
             builder.Append(";");
             builder.Append(UniqueToken);
 
+            builder.Append(";");
+            builder.Append(ContextProject);
+
             builder.Append("]");
 
-            return builder.ToString();
+            return HttpUtility.UrlEncode(builder.ToString());
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyNodeId.cs]: Can you not pull in System.Web, and instead use https://msdn.microsoft.com/en-us/library/system.net.webutility(v=vs.110).aspx.

**Comment by davkean** [comment]
Can you help me understand the commit descriptions, ie "Temporary fixes for projects hierarchy nodes"? They don't contain enough information for me to look back and understand the problem you are solving.

**Comment by davkean** [comment]
You've also seem to made the same commit with the same changes multiple times - not sure how that happened.

**Comment by abpiskunov** [comment]
@davkean something was messed up when i did rebase.. i did commits only once in parent  branch and then this 3rd level branch i rebased with master and commits got duplicated. Anyway - i am following with comments and squishing my commits into one  - all changes are related to the problem i mentioned in the PR description and temporary commits i made just to be able to work on other machine with newer build , it got kind of messy i agree :)

**Comment by abpiskunov** [comment]
Followed up with comments, squashed ambiguous commits and created a new PR: https://github.com/dotnet/roslyn-project-system/pull/1062 . Closing this one

**Pull request closed without merging**